Spark Submit Error on EMR

When running `spark-submit` in EMR we see the following error:

```
```Exception in thread "main" java.lang.NoSuchMethodError: shapeless.Witness$.mkWitness(Ljava/lang/Object;)Lshapeless/Witness;
    at com.azavea.rf.batch.ingest.model.IngestDefinition$anon$lazy$macro$35$1.inst$macro$25$lzycompute(IngestDefinition.scala:12)
    at com.azavea.rf.batch.ingest.model.IngestDefinition$anon$lazy$macro$35$1.inst$macro$25(IngestDefinition.scala:12)
    at com.azavea.rf.batch.ingest.model.IngestDefinition$.<init>(IngestDefinition.scala:12)
    at com.azavea.rf.batch.ingest.model.IngestDefinition$.<clinit>(IngestDefinition.scala)
    at com.azavea.rf.batch.ingest.Ingest$.main(Ingest.scala:292)
    at com.azavea.rf.batch.ingest.Ingest.main(Ingest.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
    at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
    at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
    at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
    at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

To test/verify that the issue is not happening locally, you can try the following command after running an assembly on the batch subproject.

```
docker-compose -f docker-compose.spark.yml run spark-driver \
  --driver-memory 2G \
  --class com.azavea.rf.batch.ingest.Ingest \
  /opt/rf/jars/rf-batch.jar \
  -j s3://rasterfoundry-development-data-us-east-1/ingest-definitions/49780d91-33ca-4467-be13-d326fefa338d.json
```

Starting a similar job in EMR will result in the error (taken from the Yarn resource manager logs):
```
tion$anon$lazy$macro$35$1.inst$macro$25(IngestDefinition.scala:12)
	at com.azavea.rf.batch.ingest.model.IngestDefinition$.<init>(IngestDefinition.scala:12)
	at com.azavea.rf.batch.ingest.model.IngestDefinition$.<clinit>(IngestDefinition.scala)
	at com.azavea.rf.batch.ingest.Ingest$.main(Ingest.scala:292)
	at com.azavea.rf.batch.ingest.Ingest.main(Ingest.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:637)
17/04/20 19:53:02 INFO ApplicationMaster: Final app status: FAILED, exitCode: 15, (reason: User class threw exception: java.lang.NoSuchMethodError: shapeless.Witness$.mkWitness(Ljava/lang/Object;)Lshapeless/Witness;)
17/04/20 19:53:02 ERROR ApplicationMaster: Uncaught exception: 
org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:194)
	at org.apache.spark.deploy.yarn.ApplicationMaster.runDriver(ApplicationMaster.scala:401)
	at org.apache.spark.deploy.yarn.ApplicationMaster.run(ApplicationMaster.scala:254)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$main$1.apply$mcV$sp(ApplicationMaster.scala:766)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:67)
	at org.apache.spark.deploy.SparkHadoopUtil$$anon$1.run(SparkHadoopUtil.scala:66)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1698)
	at org.apache.spark.deploy.SparkHadoopUtil.runAsSparkUser(SparkHadoopUtil.scala:66)
	at org.apache.spark.deploy.yarn.ApplicationMaster$.main(ApplicationMaster.scala:764)
	at org.apache.spark.deploy.yarn.ApplicationMaster.main(ApplicationMaster.scala)
Caused by: java.util.concurrent.ExecutionException: Boxed Error
	at scala.concurrent.impl.Promise$.resolver(Promise.scala:55)
	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:47)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:244)
	at scala.concurrent.Promise$class.tryFailure(Promise.scala:112)
	at scala.concurrent.impl.Promise$DefaultPromise.tryFailure(Promise.scala:153)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:655)
Caused by: java.lang.NoSuchMethodError: shapeless.Witness$.mkWitness(Ljava/lang/Object;)Lshapeless/Witness;
	at com.azavea.rf.batch.ingest.model.IngestDefinition$anon$lazy$macro$35$1.inst$macro$25$lzycompute(IngestDefinition.scala:12)
	at com.azavea.rf.batch.ingest.model.IngestDefinition$anon$lazy$macro$35$1.inst$macro$25(IngestDefinition.scala:12)
	at com.azavea.rf.batch.ingest.model.IngestDefinition$.<init>(IngestDefinition.scala:12)
	at com.azavea.rf.batch.ingest.model.IngestDefinition$.<clinit>(IngestDefinition.scala)
	at com.azavea.rf.batch.ingest.Ingest$.main(Ingest.scala:292)
	at com.azavea.rf.batch.ingest.Ingest.main(Ingest.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:637)
17/04/20 19:53:02 INFO ApplicationMaster: Unregistering ApplicationMaster with FAILED (diag message: User class threw exception: java.lang.NoSuchMethodError: shapeless.Witness$.mkWitness(Ljava/lang/Object;)Lshapeless/Witness;)
17/04/20 19:53:02 INFO ApplicationMaster: Deleting staging directory hdfs://ip-10-0-127-121.ec2.internal:8020/user/hadoop/.sparkStaging/application_1492701974127_0015
17/04/20 19:53:02 INFO ShutdownHookManager: Shutdown hook called
```

--- a/app-backend/build.sbt
+++ b/app-backend/build.sbt
@@ -175,6 +175,10 @@ lazy val batch = Project("batch", file("batch"))
       Dependencies.scopt
     )
   })
+  .settings(assemblyShadeRules in assembly := Seq(
+      ShadeRule.rename("shapeless.**" -> "com.azavea.shaded.shapeless.@1").inAll
+    )
+  )
 
 import io.gatling.sbt.GatlingPlugin
 lazy val tile = Project("tile", file("tile"))

--- a/app-backend/project/Version.scala
+++ b/app-backend/project/Version.scala
@@ -20,7 +20,7 @@ object Version {
   val akkaHttpCors       = "0.1.11"
   val akkaSlf4j          = "2.4.11"
   val ammoniteOps        = "0.7.0"
-  val spark              = "2.0.0"
+  val spark              = "2.1.0"
   val commonsIO          = "2.5"
   val scopt              = "3.5.0"
   val scaffeine          = "2.0.0"

--- a/app-backend/project/build.properties
+++ b/app-backend/project/build.properties
@@ -1,1 +1,1 @@
-sbt.version=0.13.13
+sbt.version=0.13.15

--- a/app-backend/project/plugins.sbt
+++ b/app-backend/project/plugins.sbt
@@ -9,7 +9,7 @@ addSbtPlugin("me.lessis" % "bintray-sbt" % "0.3.0")
 
 addSbtPlugin("com.timushev.sbt" % "sbt-updates" % "0.1.10")
 
-addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.14.3")
+addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.14.4")
 
 addSbtPlugin("io.spray" % "sbt-revolver" % "0.8.0")
 

